### PR TITLE
[compiler-rt][rtsan] Use sanitizer internal allocator during rtsan init to avoid segfault in dlsym

### PR DESCRIPTION
**Pull request by davidtrevelyan**
Follows https://github.com/llvm/llvm-project/pull/98268 with a fix for a segfault during preinit on `ubuntu:20.04` environments. Previously, `rtsan` was not handling the situation where `dlsym` calls `calloc` during the interceptors initialization, resulting in a call to a function at a null address. 

@cjappl and I took inspiration from the solution in `nsan`, but we re-used the sanitizer internal allocator instead of our own static buffer. This PR also re-enables the existing non-instrumented `rtsan` tests for `x86_64` and `arm64` architectures.

### CODE REVIEW

**Comment by davidtrevelyan** [comment]
Tagging interested parties: @vitalybuka @cjappl @chapuni - many thanks in advance for any reviews.

**Review by chapuni — APPROVED** [review]
This works for me.
Please wait for other reviews.

**Comment by davidtrevelyan** [comment]
Polite ping FAO @vitalybuka and @MaskRay - your oversight on this PR would be much appreciated, @cjappl and I are just waiting for this and https://github.com/llvm/llvm-project/pull/99437 (also approved) to go in before we submit our next PR for rtsan's frontend in clang. 

Just wanted to also leave a reminder that I don't have commit access, so I will need someone to do the merge for me if there are no further requested changes. 

Many thanks in advance for your time!

File: compiler-rt/lib/rtsan/rtsan_interceptors.cpp
Comment (MaskRay) [inline compiler-rt/lib/rtsan/rtsan_interceptors.cpp:246]: We should define `UseImpl` and use `DlsymAlloc::Use`. See #98415

Comment (cjappl) [reply compiler-rt/lib/rtsan/rtsan_interceptors.cpp]: Fixed! Thanks

**Review by MaskRay — APPROVED** [review]

**Comment by cjappl** [comment]
Also, neither of the main authors of this patch have review access. If someone sees this after it has an acceptable number of approvals and wants to merge it, that would be great! Thanks

**Comment by MaskRay** [comment]
I was waiting for possible response from regular compiler-rt sanitizers folks:) I think this is good to merge in the absence of their response.

**Comment by cjappl** [comment]
Thanks @MaskRay ! We were unsure of what to do when it looked like there was some approval, but we didn't have merge rights. Thanks for your help as always

**Comment by cjappl** [comment]
Having an error on android, which I thought was previously removed but may have been caught in a revert conflict.

```
OSError: [Errno 8] Exec format error: '/var/lib/buildbot/sanitizer-buildbot6/sanitizer-x86_64-linux-android/build/compiler_rt_build_android_aarch64/lib/rtsan/tests/./Rtsan-aarch64-NoInstTest'
```

https://github.com/llvm/llvm-project/pull/99964

**Pull request merged**
